Watchdog (docker container moniotring module)

adonagy opened the pull request:
Adding the watchdog/monitoring app. The whole dir has been formatted using `cargo fmt`.

TODOs:

- [x] add a README
- [x] Use an environment var to set the required image tag instead of prog. arg. (e.g. TEZEDGE_NODE_IMAGE_TAG) 
- [x] Push a notification after the images have been updated

bkontur commented:
@adonagy 
Ado, move watchdog do subdir: `apps/watchdog` according to https://github.com/simplestaking/tezedge/issues/330

adonagy commented:
@bkontur @vlad9486 
I rebased and updated the dependencies, also bumped tokio to version 1.1. 
As a note, we should watch the newest release of the shiplift crate, cargo is getting the crate from the master branch (they latest release is quite old and we need features from master). Their update to tokio 1.1 caused some version conflict in the watchdog app, which is now fixed.

bkontur reviewed (commented):
@adonagy 
Ado, please rebase on develop

bkontur commented on apps/watchdog/src/configuration.rs:
```diff
@@ -0,0 +1,209 @@
+
+fn validate_required_args(args: &clap::ArgMatches) {
+    validate_required_arg(args, "image-monitor-interval");
+    // validate_required_arg(args, "resource-monitor-interval");
```
@adonagy 
please, remove unused code

bkontur commented on apps/watchdog/src/deploy_with_compose.rs:
```diff
@@ -0,0 +1,158 @@
+use crate::image::{Explorer, OcamlDebugger, Sandbox, TezedgeDebugger, WatchdogContainer};
+use crate::node::{OcamlNode, TezedgeNode};
+
+// TODO: use external docker-compose for now, should we manage the images/containers directly?
```
@adonagy 
what about this TODO? please create jira for that or remove comment, if not needed

bkontur commented on apps/watchdog/src/deploy_with_compose.rs:
```diff
@@ -0,0 +1,158 @@
+    start_with_compose(compose_file_path, Explorer::NAME, "explorer");
+    start_with_compose(compose_file_path, TezedgeDebugger::NAME, "tezedge-debugger");
+    // debugger healthcheck
+    while reqwest::get("http://localhost:17732/v2/log").await.is_err() {
```
@adonagy 
please, create const for `http://localhost:17732/v2/log` (maybe it should be better as part of configuration, but at least const is better)

bkontur replied:
@adonagy 
I see below there is also a constant:
`pub const TEZEDGE_DEBUGGER_PORT: u16 = 17732;`
I would put that to one place, maybe just struct with constants

bkontur commented on apps/watchdog/src/deploy_with_compose.rs:
```diff
@@ -0,0 +1,158 @@
+
+    start_with_compose(compose_file_path, TezedgeNode::NAME, "tezedge-node");
+    // node healthcheck
+    while reqwest::get("http://localhost:18732/chains/main/blocks/head/header")
```
@adonagy 
also as const (better cfg resolved from compose :D, please as const for the begining)

bkontur commented on apps/watchdog/src/deploy_with_compose.rs:
```diff
@@ -0,0 +1,158 @@
+        .await
+        .is_err()
+    {
+        sleep(Duration::from_millis(1000)).await;
```
@adonagy 
please, create const for this sleep

bkontur commented on apps/watchdog/src/display_info.rs:
```diff
@@ -0,0 +1,387 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+#![forbid(unsafe_code)]
```
@adonagy 
I guess, this is needed just in lib.rs or mod.rs

bkontur commented on apps/watchdog/src/display_info.rs:
```diff
@@ -0,0 +1,387 @@
+}
+
+#[derive(Serialize)]
+pub struct MemoryData {
```
@adonagy 
I think this could be simplified/unified, ocaml and tezedge looks the same:
```
main_process: ProcessMemoryStats,
sub_processes: Vec<ProcessMemoryStats>
```

less splitting things for ocaml vs tezedge is better

bkontur commented on apps/watchdog/src/monitors/deploy.rs:
```diff
@@ -0,0 +1,170 @@
+            let node_updated = self.changed::<TezedgeNode>().await?;
+            let debugger_updated = self.changed::<TezedgeDebugger>().await?;
+            let explorer_updated = self.changed::<Explorer>().await?;
+            // TODO: here restart everything,
```
@adonagy 
TODO, please create jira or remove comment, if not needed

bkontur commented on apps/watchdog/src/monitors/mod.rs:
```diff
@@ -0,0 +1,148 @@
+pub mod info;
+pub mod resource;
+
+// TODO: get this info from docker (shiplift needs to implement docker volume inspect)
```
@adonagy 
TODO, please create jira or remove if not needed

bkontur reviewed (changes requested):
@adonagy 
please see PR comments and rebase on develop

adonagy commented:
Closing this PR. Continues in #528
